doc/design.md

```diff
@@ -0,0 +1,922 @@
+Instead `DatasetIndex` builds the index (and does the validation) on demand when constructed.
+
+##### Rationale
+
+I tried to *maintain* consistent indices like spectrum numbers using code in `Indexing::IndexInfo`, but it turned out to be really complicated when integrating it into `API::MatrixWorkspace`.
+
+
+### <a name="design-components-units"></a>Units
+
+Internally the prototype uses `boost::units`, but we require run-time units.
+It is therefore wrapped in a class that simply stores an unique ID for every unit.
+- Arithmetic operations branch off using the ID, relying on `boost::units` for the actual work.
+- Probably need to add support for units scales.
+
+Very little of this is implemented in the prototype apart from the most basic examples.
+This will probably require a not insignificant amount of work to get right to make sure we capture all the subtleties.
+See for example [An issue with distributions and dimensionless units](https://github.com/mantidproject/documents/blob/fix-divide-distribution/Design/DistributionsAndDimensionlessData.md) for a recent issue relating to this in Mantid.
+
+
+### <a name="design-components-exceptions"></a>Exceptions
+
+The prototype currently simply throws `std::runtime_error` in most cases.
+We should most likely add a clear range of exception types that provide useful information such as indices about the failure.
+
+
+## <a name="design-details"></a>Design details
+
+
+### <a name="design-details-dependencies"></a>Dependencies
+
+- `boost::units` for unit handling.
+- `boost::container` for `small_vector`, which is used as an optimization in a couple of places.
+- `boost::mpl` to help with some meta programming.
+- `Eigen` for vectors and rotations, in particular for detector positions.
+- `gsl` is used for `gsl::index` and `gsl::span`.
+  This could easily be replaced by a stand alone implementation, in particular since `span` is part of C++20.
+- `range-v3` has been used for implementing some multi-column sort operations but replacing it by a stand-alone implementation should be no trouble.
+- GoogleTest for unit testing.
+- GoogleBenchmark for benchmarking.
+
+The prototype has been implemented using C++14, and this is sufficient.
+Nevertheless, having C++17 available would be an advantage, particularly since a moderate amount of template-meta programming is required.
+Given the long time scale, it could be expected that the Mantid project is ready to move to C++17 by the time `Dataset` might be included in the main repository.
+
+Note that `Dataset` or rather `Variable` will be able to hold data from libraries that are *not* dependencies of this library.
+It is sufficient to forward-declare types when adding a tag for such a variable.
+*Example:
+We can store `API::Run` in `Dataset`, but the dataset library does not need to depend on `Mantid::API`.*
+
+It is currently unclear whether this library should be kept as a separate repository or be included as a new module in the Mantid repository.
```

Comment:
I prefer separation from day-1 if we can reasonably achieve it. 1. Less change of it being corrupted/bound tightly into the rest of Mantid. 2. Potential for outside developers to take advantage without dragging in the rest of Mantid.

Reply:
I added a comment to require shipping it separately if included in Mantid.